### ISSUE TITLE
Running Docker processes as root

### ISSUE THREAD

**Issue by erewok** [phadej/docker-haskell-example#2]
Hello @phadej! I was looking for some multi-stage build Dockerfiles for Haskell projects and I happened upon this one, which has some great tips. I think one currently agreed-upon Docker best practice is missing, though, from your Dockerfile: don't run processes as root. It may be worth changing here if people are finding your blog-post and copying the code in there? I could probably submit a PR to this repo if you think it's worth it.

**Comment by phadej** [comment]
Do you mean the final app?

Yes, I guess user can be created at the very last step.
I however don't know what are the *best practices for creating users in Dockerfiles*. I'd like to have a reference for that.

**Comment by erewok** [comment]
> Do you mean the final app?

Yes, exactly, the final executable. We do something like this for our `debian`-based images somewhere _after_ any system things (`apt`, etc.) and before the `ENTRYPOINT`:

```
RUN  adduser --disabled-password --gecos "" -u 999 some_user
USER some_user
```

It requires some care to make sure the user has `+x` on the directory hierarchy where the executable lives and they can read and execute it, but that's pretty much it.

Not doing the above, by the way, is one of the largest and most common security issues in docker containers.

**Comment by erewok** [comment]
For references, in a few minutes of searching, I found the following, among other articles: 

- https://medium.com/@mccode/processes-in-containers-should-not-run-as-root-2feae3f0df3b
- https://medium.com/better-programming/docker-best-practices-and-anti-patterns-e7cbccba4f19

**Comment by phadej** [comment]
I'm not convinced.

If I can `docker run` on machine it's virtually the same as having root access anyway.

Second post says

> A Docker container running as root has full control of the host system.

Isn't that just wrong? E.g. GitHub actions allows to run containers where root privileges aren't dropped. Does it mean we all have access to GitHub Actions hosts. I don't think so.

I'm also more suspicious as the post suggest doing

---

I'm not opposed to the change, but these reference don't convince me that I do things for nothing.

---

The first post actually says

> Finally, you can create a user on the host, and pass its uid to Docker when starting the container. For example, revisiting the original example Dockerfile:

```
FROM debian:stretch
CMD ["cat", "/tmp/secrets.txt"]
```

> I can run this container with or without a user id parameter and see the different results (user id 1001 is my own account on this server):

```
$ docker run --user 1001 -v /root/secrets.txt:/tmp/secrets.txt <img>
cat: /tmp/secrets.txt: Permission denied
```

```
$ docker run -v /root/secrets.txt:/tmp/secrets.txt <img>
top secret stuff
```

> This works and does the same thing as creating a user in the Dockerfile, but it requires the user to optionally run the container securely. 

And I think **that is correct way** to run arbitrary docker images (rather then validating that they have proper setup inside them).

**Comment by erewok** [comment]
Well, I certainly won't try to convince you but it is a widely-agreed upon best practice to not run processes in containers as root. 

For instance, your readers may be running Docker containers in a variety of environments, possibly on cloud VMs or in Kubernetes clusters. I watched a talk last month at Kubecon where the speaker talked about integrating seccomp into Docker many years ago and later commented that in spite of all of these security efforts their largest hurdle is still trying to get people to stop running processes as root. This topic comes up frequently when discussing securing docker containers in kubernetes.

Of course, similar to what you're saying here, Kubernetes also allows users to specify UID and GID when running pods.

**Comment by phadej** [comment]
> Of course, similar to what you're saying here, Kubernetes also allows users to specify UID and GID when running pods.

And why that is not recommended? What are the drawbacks of that?

**Comment by erewok** [comment]
Oh, apologies if I implied it wasn't. It certainly is. For instance, here are the [Azure docs that mention that](https://docs.microsoft.com/en-us/azure/aks/developer-best-practices-pod-security#secure-pod-access-to-resources).
